### DISCUSSION TITLE
One-dimensional form of the incompressible Navier-Stokes equations

### DISCUSSION THREAD

**Discussion by vl-dud** [lululxvi/deepxde#1231] [Q&A] [answered]
Hello, I'm interested in using deepxde to model coronary vessels with stenosis. I am considering a one-dimensional form of the incompressible Navier-Stokes equations. PDEs takes the form:

![1](https://user-images.githubusercontent.com/60846135/229499814-ebe2d597-a001-4bc1-b28f-11bbdb50f8db.png)

where A(x,t), u(x,t) and p(x,t) represent the cross-sectional area, the velocity and pressure, respectively. They are output values. Other variables (A0, Pext, etc) will be regarded as fixed values.
My code:
```
import deepxde as dde
import math
from deepxde.backend import tf


rho = 1060. # density (kg/m^3)
A0 = 9.6402913e-06 # vessel’s sectional area at equilibrium (m^2)
p0 = 45000. # initial pressure
KR = -0.241902634 # friction parameter
p_ext = 10000. # external pressure
alpha = 1.1
beta = 32831955.


def pde(x, y):
    u = y[:, 0:1]
    A = y[:, 1:2]
    p = y[:, 2:3]

    du_x = dde.grad.jacobian(y, x, i=0, j=0)
    du_t = dde.grad.jacobian(y, x, i=0, j=1)
    dA_x = dde.grad.jacobian(y, x, i=1, j=0)
    dA_t = dde.grad.jacobian(y, x, i=1, j=1)
    dp_x = dde.grad.jacobian(y, x, i=2, j=0)

    continuity_equation = dA_t + dA_x * u + du_x * A
    momentum_equation = du_t + alpha * u * du_x + (u / A) * (alpha - 1) * (dA_x * u + du_x * A) + dp_x / rho - KR * u / A
    pressure_equation = p_ext + beta * (tf.sqrt(tf.abs(A)) - math.sqrt(A0)) - p
    return [continuity_equation, momentum_equation, pressure_equation]


# vessel length is 0.11 m
space_domain = dde.geometry.Interval(0, 0.11)
time_domain = dde.geometry.TimeDomain(0, 10)
geomtime = dde.geometry.GeometryXTime(space_domain, time_domain)

# some initial conditions
ic1 = dde.icbc.IC(geomtime, lambda x: A0, lambda x, on_initial: on_initial, component=1)
ic2 = dde.icbc.IC(geomtime, lambda x: p0, lambda x, on_initial: on_initial, component=2)

data = dde.data.TimePDE(
    geomtime,
    pde,
    [ic1, ic2],
    num_domain=2000,
    num_initial=500,
    num_test=200
)

layer_size = [2] + [50] * 6 + [3]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_size, activation, initializer)


def output_transform(x, y):
    _x = x[:, 0]
    # stenosis in a center of the vessel, length is 0.01 m
    mask = tf.logical_and(_x >= 0.05, _x <= 0.06)
    multiplier = 0.33
    # decrease A
    y_new = tf.where(mask, y[:, 1:2] * multiplier, y[:, 1:2])
    return tf.concat([y[:, 0:1], y_new, y[:, 2:]], axis=1)


net.apply_output_transform(output_transform)
model = dde.Model(data, net)
model.compile("adam", lr=0.001)
losshistory, train_state = model.train(iterations=10000)
```

I have some questions.

1. Is `net.apply_output_transform` the right way to define stenosis?
2. Do I need to generate numerous geometries (GeometryXTime objects) to train the model to work with various vessel lengths and stenosis? I can't find similar examples in the FAQ.

I would greatly appreciate any advice or recommendations as well.

**Comment by lululxvi** [comment]
1. Use `_x = x[:, 0:1]`
2. Can you be more specific? You try to solve multiple problems?

**Reply by vl-dud** [reply]
1. Thank you
2. I am trying to train the model so that it can work with different types of stenosis. Stenosis is determined by a reduction in the cross-sectional area, which is represented by the output value A. It may occur at different positions along the X-axis (now exactly in the middle) and can come in varying sizes (33% in my code). Perhaps the stenosis should be set as initial conditions instead of `output_transform`? But it is unclear how to train the model to handle such cases and how it will perform with real-world examples.

**Reply by lululxvi** [reply, marked as answer]
Check DeepONet.
